src/caf/toolkit/cost_utils.py

```diff
@@ -3,25 +3,422 @@
+
+    # Ideas
+    units: str = "km"
+
+    @pydantic.root_validator
+    def check_df_col_names(cls, values):  # pylint:disable=no-self-argument
+        """Check the given columns are in the given dataframe."""
+        # init
+        col_names = ["min_col", "max_col", "avg_col", "trips_col"]
+        cols = {k: v for k, v in values.items() if k in col_names}
+        df = values.get("df")
+
+        # Check columns are in df
+        err_cols = {}
+        for col_name, col_val in cols.items():
+            if col_val not in df:
+                err_cols.update({col_name: col_val})
+
+        # Throw error if missing columns found
+        if err_cols != dict():
+            raise ValueError(
+                "Not all the given column names exist in the given df. "
+                f"The following columns are missing:{err_cols}\n"
+                f"With the following in the Df: {df.columns}"
+            )
+
+        # Tidy up df
+        values["df"] = pd_utils.reindex_cols(df, cols.values())
+        return values
+
+    def __len__(self):
+        """Get the number of bins in this cost distribution."""
+        return len(self.bin_edges) - 1
+
+    def __eq__(self, other):
+        """Check if two items are the same."""
+        if not isinstance(other, CostDistribution):
+            return False
+        return (self.df == other.df).values.all()
+
+    def copy(self) -> CostDistribution:
+        """Create a copy of this instance."""
+        return copy.copy(self)
+
+    @property
+    def min_vals(self) -> np.ndarray:
+        """Minimum values of the cost distribution bin edges."""
+        return self.df[self.min_col].values
+
+    @property
+    def max_vals(self) -> np.ndarray:
+        """Maximum values of the cost distribution in edges."""
+        return self.df[self.max_col].values
+
+    @property
+    def bin_edges(self) -> np.ndarray:
+        """Bin edges for the cost distribution."""
+        return np.append(self.min_vals, self.max_vals[-1])
+
+    @property
+    def n_bins(self) -> int:
+        """Bin edges for the cost distribution."""
+        return len(self)
+
+    @property
+    def avg_vals(self) -> np.ndarray:
+        """Average values for each of the cost distribution bins."""
+        return self.df[self.avg_col].values
+
+    @property
+    def trip_vals(self) -> np.ndarray:
+        """Trip values for each of the cost distribution bins."""
+        return self.df[self.trips_col].values
+
+    @property
+    def band_share_vals(self) -> np.ndarray:
+        """Band share values for each of the cost distribution bins."""
+        trip_vals = self.trip_vals
+        return trip_vals / np.sum(trip_vals)
+
+    @staticmethod
+    def from_data(
+        matrix: np.ndarray,
+        cost_matrix: np.ndarray,
+        min_bounds: Optional[list[float] | np.ndarray] = None,
+        max_bounds: Optional[list[float] | np.ndarray] = None,
+        bin_edges: Optional[list[float] | np.ndarray] = None,
```

Comment:
If you're using `np.histogram` then this can technically be an int for number of bins, with a few changes to this method, is that functionality that is worth enabling?

Reply:
I hadn't thought of it in that way, but it's a valid point.

I think it's something worth considering, but maybe out of scope for this commit. I'll add this as an issue for us to come back to later.

Reply:
Issue added at #64